Price group: hiding plans according to the price group.

southp opened the pull request:
#### Changes proposed in this Pull Request
**Description update: May. 7th.**

*This is part of the Global Pricing Optimization project. More context can be found in p2-pau2Xa-1de.*

This PR implements the plan hiding logic based on the marketing price group, for hiding plans accordingly in the signup flow, the site creation flow, and viewing `/plans` as a logged-in user.

#### Dependency

* https://github.com/Automattic/wp-calypso/pull/41871
* #41670 (shipped)
* code-D43026
* code-D42721 (shipped)

#### Testing instructions

Please refer to the test section of code-D42721 to see how to enforce a user into either test pricing group.

1. Enforce a test account to the test pricing group 1.
    1. Go through the sign up flow and stop at the plans step.
    1. While there shouldn't be any visible difference, `getState().signup.dependencyStore` should include a `marketing_price_group` property, with value `price_2020_q2_test_1`.
    1. Stay logged-in, go through the site creation flow and stop at the plans step. 
    1. Stay logged-in, go to `/plans`.     
1. Enforce a test account to the test pricing group 2.
    1. Go through the sign up flow and stop at the plans step.
    1. The Personal plan should be hidden.
    1. Stay logged-in, go through the site creation flow and stop at the plans step. The Personal plan should be hidden.
    1. Stay logged-in, go to `/plans`. The Personal plan should be hidden.
1. Enforce a test account to the test pricing group 3.
    1. Go through the sign up flow and stop at the plans step.
    1. The Personal plan and the Premium plan should be hidden.
    1. Stay logged-in, go through the site creation flow and stop at the plans step. The Personal plan and the Premium plan should be hidden.
    1. Stay logged-in, go to `/plans`. The Personal plan and the Premium plan should be hidden.

southp committed:
Hide the Personal & Premium plans according to the price group value.

niranjan-uma-shankar commented on client/state/selectors/get-current-user-marketing-price-group.js:
```diff
@@ -0,0 +1,16 @@
+export default ( state ) => {
+	const currentUser = getCurrentUser( state );
+
+	return ( currentUser && currentUser.meta.marketing_price_group ) || null;
```
Small thing, but we could use [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) here:

> return currentUser?.meta.marketing_price_group
or the following if you want `null` returned instead of `undefined`
> return currentUser?.meta.marketing_price_group || null

southp replied:
Thanks for the tip! I have updated the code in fa49661.

niranjan-uma-shankar reviewed (commented):
Looking good @southp. Found a minor issue.
I tested with the backend patch. When I visit the plans page just after signup, or right after "Add new site" flow is complete, then the plans page alignment is not proper. If I reload the page, then the alignment fixes itself. I would guess this is not a blocker, but just bringing to your notice. 
Check screen recording here - https://d.pr/v/JlTXPX 

Also, could you please rebase this branch so that I can test the changes done in https://github.com/Automattic/wp-calypso/pull/41526?

niranjan-uma-shankar reviewed (commented):
I tested a bit more and I found a few places where we explicitly mention the Personal and Premium plan:

<img width="1089" alt="Screenshot 2020-05-04 at 6 28 05 PM" src="https://user-images.githubusercontent.com/1269602/80971297-2b10f380-8e3a-11ea-8a07-af82949ff2e5.png">

<img width="1094" alt="Screenshot 2020-05-04 at 6 29 05 PM" src="https://user-images.githubusercontent.com/1269602/80971346-42e87780-8e3a-11ea-9550-346bfac46506.png">

<img width="755" alt="Screenshot 2020-05-04 at 7 03 53 PM" src="https://user-images.githubusercontent.com/1269602/80971368-4b40b280-8e3a-11ea-90af-dc62aeeba951.png">

<img width="1535" alt="Screenshot 2020-05-04 at 7 02 05 PM" src="https://user-images.githubusercontent.com/1269602/80971436-66132700-8e3a-11ea-8df0-ac23b8cdabc4.png">

<img width="748" alt="Screenshot 2020-05-04 at 6 57 16 PM" src="https://user-images.githubusercontent.com/1269602/80971461-71665280-8e3a-11ea-81c2-6c43946fb0f1.png">

There are several more places where we show these upgrade nudges. I suggest the following search terms to find references in code:

https://github.com/Automattic/wp-calypso/search?q=%22WordPress.com+Premium%22&unscoped_q=%22WordPress.com+Premium%22

https://github.com/Automattic/wp-calypso/search?q=%22WordPress.com+Personal%22&unscoped_q=%22WordPress.com+Personal%22

We also refer to these plans in other ways, e.g. Premium plan, WordPress.com Premium, Upgrade to Premium etc. so we need to find all those references. For eg. users from Canada and US will see the following card in the Marketing page:

<img width="606" alt="Screenshot 2020-05-04 at 7 22 20 PM" src="https://user-images.githubusercontent.com/1269602/80973341-dcb12400-8e3c-11ea-845e-a2ebc2efd6b9.png">

Some other uses of Premium/Personal variations:

<img width="604" alt="Screenshot 2020-05-04 at 7 15 37 PM" src="https://user-images.githubusercontent.com/1269602/80973447-0407f100-8e3d-11ea-9784-a1a482b896fe.png">

<img width="661" alt="Screenshot 2020-05-04 at 7 11 45 PM" src="https://user-images.githubusercontent.com/1269602/80973513-197d1b00-8e3d-11ea-93f1-2e0e74b566a4.png">

southp committed:
Derive `hidePersonalPlan` and `hidePremiumPlan` according to
`marketingPriceGroup` in the plan step of the sign-up flow and the site creation flow.

southp committed:
Use optional chainning to simplify the implementation of
`getCurrentUserMarketingPriceGroup()`.

southp commented:
@niranjan-uma-shankar Now that all the dependent branches and the backend diff have been deployed, this PR should be ready for a revisit again. We aim at shipping this test within this week, so it would be wonderful if you can drive by when you have some time today :)

niranjan-uma-shankar commented on client/lib/signup/step-actions/index.js:
```diff
@@ -450,7 +450,10 @@ export function createAccount(
-				callback( undefined, pick( response, [ 'username', 'bearer_token' ] ) );
+				callback(
+					undefined,
+					pick( response, [ 'username', 'bearer_token', 'marketing_price_group' ] )
```
Does the usersSocialNew endpoint return the marketing_price_group value? Also, like done in the else block, it's safer to assign to a const variable:

> const marketing_price_group = ( response && response.marketing_price_group ) || '';

simison replied:
You can simplify tad bit to:
`response?.marketing_price_group ?? ''`

niranjan-uma-shankar replied:
Yes, optional chaining for the win. Looks cleaner.

southp replied:
Cool tip! Added in d48c50b.

niranjan-uma-shankar commented:
@southp I tested and this works well. Left a comment on social signup above, but otherwise looking good. I can do a final run tomorrow after the social signup thing is fixed.

southp committed:
Add `marketing_price_group` to the signup depdendency store.

southp committed:
Use optional chainning to simplify how `marketing_price_group` is
derived.

southp commented:
@niranjan-uma-shankar Thanks for finding it. It's definitely something I overlooked. The patch to update the social signup endpoint is opened as code-D43026.

Also, to address the code-wise issue, I refactored `createAccount()` in https://github.com/Automattic/wp-calypso/pull/41871 and then implemented the requested change here again.

niranjan-uma-shankar reviewed (commented):
Do we want to add a primary button(pink colour) for the plan options in CAD and AUD? 

In signup flow:

<img width="889" alt="Screenshot 2020-05-07 at 10 36 33 AM" src="https://user-images.githubusercontent.com/1269602/81256714-f1650600-904e-11ea-8cd8-376f1c0dc8aa.png">

In /plans page:

<img width="818" alt="Screenshot 2020-05-07 at 10 37 28 AM" src="https://user-images.githubusercontent.com/1269602/81256726-f629ba00-904e-11ea-8947-3fb8c6259c75.png">

southp commented:
@niranjan-uma-shankar Good catch. Since this PR and the related diff is already quite involving, I think it would be better to be fixed in a separate PR. Does that make sense to you?

niranjan-uma-shankar commented:
Makes sense

niranjan-uma-shankar reviewed (commented):
Noting here an issue which is an edge case, and I don't think we should try fixing it for this test. This is just for the record so that if we get that "lucky" user who faces the issue, we'll know the cause.

As you know, we have a domain-plan-swap test running but that's only for EN-US. We get  geo location from wpcom cookie. The cookie value stays even if the user moves to a different location i.e its value is not overwritten most of the times, except in some edge cases. Now, if a user satisfies the following conditions, there's a scenario in which they'll see an empty plans page:
1. User creates their account from Canada or Australia, and gets assigned to the pricing test. So, Personal & Premium plan is hidden.
2. The user later deletes their browser cookies for wpcom. Several ways why this could happen - either they flushed their browser, or used an incognito window for wpcom so cookies aren't stored, A more common cause could be that they later use a different device from the one they used to create the account.
3. The user logs in to wpcom later, but this time from a US IP address. Either they use a new device like mobile or tab from a US IP, or they use the same device but with cookies previously cleared. The wpcom cookie will now store "US" as the location. Their currency will still be CAD or AUD since that doesn't change.
4. If they click "Add new site", they will be assigned to the swap test. Say they got assigned to the variant, so they see the steps swapped. 
5. If they select a free plan -> custom domain -> click Yes on the upsell offer -> they will see an empty plans page(we show a modified plans page for those who click Yes on the upsell offer page with just the Personal & Premium plan, but this will be empty since both plans are removed).

Screen recording - https://d.pr/v/RB6ATS

Again, a lot of things have to align for this and it is an edge case.

niranjan-uma-shankar reviewed (approved):
LGTM :shipit:
